Switch from json to yaml

razvan opened the pull request:
Fixes #24

maltesander committed:
Started to adapt test jsons to yaml

maltesander committed:
Adapted yaml with unit anchors readable by serde.

maltesander committed:
Test yaml works with anchors for units and dependsOn properties.

maltesander committed:
Deserializer works for automatic string to regex conversion.

maltesander committed:
WIP

maltesander committed:
WIP - merge properties works

maltesander committed:
WIP

razvan committed:
Make tests compile

maltesander committed:
WIP - added some validation

razvan committed:
wip

maltesander committed:
WIP - some refactoring in validation

maltesander committed:
WIP - started with test setup

razvan committed:
More tests.

razvan committed:
Some more tests.

razvan committed:
Update inline docs and visibility of some functions.

maltesander committed:
Added attribute "no_copy" to PropertySpec for meta data switches that only exist to be expanded.
Removed Json config files.
Added tests.

maltesander committed:
Removed warnings.

razvan committed:
Added first validation test.

razvan committed:
Even more tests

maltesander committed:
Created generic java properties writer that works with HashMap and BTreeMap.

maltesander committed:
Adapted method name.

razvan committed:
Replace as_ calls with to_

maltesander committed:
Removed println for product config.

razvan committed:
More tests for property validation.

maltesander committed:
Fixed clippy warnings.

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+        result
+    }
+
+    /// Validates the given [merged_properties] by performing the following actions:
```
Do links to own arguments work?

maltesander replied:
I dont quite understand? Can you elaborate?

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+            match (prop, value) {
+                (Some(property), Some(val)) => {
+                    let check_datatype = validation::check_datatype(&property, &name, &val);
+                    if check_datatype.is_err() {
```
I believe a if let makes more sense here so you avoid the `err().unwrap()` later

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+                    // TODO: deprecated check
+                    // TODO: allowedValues
+
+                    // value is valid, check if it matches recommended or default values
+                    // was provided by recommended value?
```
This is not a proper sentence. Something is off here (the second line mostly)

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+                    // TODO: deprecated check
+                    // TODO: allowedValues
+
+                    // value is valid, check if it matches recommended or default values
+                    // was provided by recommended value?
+                    if let Some(recommended) = &property.recommended_values {
+                        let recommended_value =
+                            property.filter_value(version, recommended.as_slice());
+                        if recommended_value == Some(val.to_string()) {
+                            result.insert(
+                                name.to_string(),
+                                PropertyValidationResult::RecommendedDefault(val.to_string()),
+                            );
+                            continue;
+                        }
+                    }
+
+                    // was provided by recommended value?
+                    if let Some(default) = &property.default_values {
+                        let default_value = property.filter_value(version, default.as_slice());
+                        if default_value == Some(val.to_string()) {
+                            result.insert(
+                                name.to_string(),
+                                PropertyValidationResult::Default(val.to_string()),
+                            );
+                            continue;
+                        }
+                    }
+
+                    result.insert(
+                        name.to_string(),
+                        PropertyValidationResult::Valid(val.to_string()),
+                    );
```
This seems complicated to me.

Why not do it differently:
- Iterate over each user-provided property and validate (i.e. iterate over `user_config` instead of over `self.config.properties` in `get_and_expand_properties`)
- Check if there are any values that are required that are missing
- Check if there are any values that are missing that have recommended defaults

This would mean changing/removing the `get_and_expand_properties` function I believe.

maltesander replied:
Decided to keep as it is.

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+                    );
+                }
+                // if required and not set -> error
+                (Some(_property), None) => {
+                    //if property.has_role_required(role) {
```
Why the commented out line? I don't really understand this case.

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+                    result.insert(
+                        name.clone(),
+                        PropertyValidationResult::Error(
+                            "".to_string(),
```
Should be a real explanation

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+                    );
+                    //}
+                }
+                // override
```
As mentioned before: Not really an override but an unknown instead.

lfrancke commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+        version: &Version,
+    ) -> Option<PropertySpec> {
+        for property_anchor in &self.config.properties {
+            if property_anchor.name_from_kind(kind) != Some(name.to_string()) {
+                continue;
+            }
```
Again, maybe I totally misunderstand what's going on here but in the `name_from_kind` method you just pick the first one that matches the correct kind?

Isn't that...wrong?

maltesander replied:
Yes but it (should) be unique. You will only have one Env / Cli kind and multiple Files (with different names though)

lfrancke commented on src/types.rs:
```diff
@@ -1,36 +1,173 @@
+        None
+    }
+
+    pub fn name_from_kind(&self, kind: &PropertyNameKind) -> Option<String> {
```
I've commented above but I don't understand what this method does. Or why.
Can you explain and/or add documentation?

lfrancke commented on src/types.rs:
```diff
@@ -61,67 +198,126 @@ impl PropertyNameKind {
+    })
+}
+
+/// This is a workaround to deserialize a String directly into a compiled regex.
```
I don't understand why we can't deserialize to Regex directly even with this comment.
Are you saying that something needs Eq, PartialOrd etc.?

Have you looked into using https://github.com/tailhook/serde-regex instead?

maltesander replied:
Not sure why it didnt work. Will try.

maltesander replied:
Serde-regex does not implement Eq/PartialOrd/JsonSchema etc which we need later on to generate crds. So we stay with our own implementation.

lfrancke commented on src/validation.rs:
```diff
@@ -1,23 +1,13 @@
 
 pub type ValidationResult<T> = Result<T, Error>;
-
+/*
```
...

lfrancke commented on src/validation.rs:
```diff
@@ -178,34 +171,34 @@ pub(crate) fn validate_property_spec(
-                }
-            }
-        }
+        // // prepare "user" data
```
...

lfrancke replied:
I did not look at this file past this point because it's very hard to see what' actually in use and what's not.

lfrancke commented on src/writer.rs:
```diff
@@ -17,19 +17,44 @@ pub enum PropertiesWriterError {
 }
 
+/// Generic method to write java properties
+/// Accepts HashMap<String, Option<String>> or BTreeMap<String, Option<String>>.
```
Should probably document what you're doing with empty values

lfrancke commented on src/writer.rs:
```diff
@@ -17,19 +17,44 @@ pub enum PropertiesWriterError {
+    for (k, v) in properties {
+        if let Some(value) = v {
+            if value.is_empty() {
+                writer.write(&k, "\"\"")?;
```
I'm not convinced this is correct but let's leave it in here for now.

maltesander replied:
We had a discussion where we basically need to support:
key="value"
key=""
key=
key
...
This requires us (unrelated to the writer) to work with <String, Option<String>> all the time.

maltesander commented on src/lib.rs:
```diff
@@ -43,305 +47,654 @@ pub enum PropertyValidationResult {
+        //let mut result_config: BTreeMap<String, PropertyValidationResult> = BTreeMap::new();
 
-        let product_version = semver_parse(version)?;
+        let product_version = semver_parse(version).unwrap();
```
We can not use "?" since the method returns no result. 

For now the version comes directly from the operators and the custom resource (which therefore should be fine as long as we stick to the enums). 

If we ever change that (away from the enums towards ranges what we kinda discussed at some point) we should definitely have a result here. I will add it for now.

What do you think? This is important since it is the only interface we have.

maltesander commented on src/types.rs:
```diff
@@ -1,36 +1,173 @@
+    }
+
+    pub fn is_version_supported(&self, product_version: &Version) -> ValidationResult<bool> {
+        Ok(semver_parse(&self.as_of_version)? <= *product_version)
```
Yes we can write a deserializer for as_of_version and the version to/from range in the default or recommended values.

maltesander commented on src/util.rs:
```diff
@@ -1,218 +1,62 @@
-                        );
-                    }
-                }
+pub(crate) fn expand_properties(
```
Added some documentation, i think we clarified the other parts, so i resolve it.

maltesander committed:
Adapted partially to PR review.
Added deprecated warning and allowed value checks.
Added/Adapted tests.
Removed most of the commented code in validation.rs.

razvan committed:
Minor doc update to StackableRegex.

maltesander committed:
Adapted comment.

maltesander committed:
Added custom StackableVersion to wrap SemVer version. Added deserializer and serializer for (Option) StackableVersion to avoid parsing the previous String versions over and over.

maltesander committed:
Removed obsolete code from validation.rs

lfrancke commented on src/lib.rs:
```diff
@@ -290,28 +318,26 @@ impl ProductConfigManager {
                         name.clone(),
                         PropertyValidationResult::Error(
-                            "".to_string(),
+                            name.to_string(),
```
nitpick but the name is already in the nested error
This should be something along the lines of `The property NAME is required but missing`

maltesander replied:
The property name is there, its the value that is missing?

lfrancke reviewed (dismissed):
One comment left, up to you whether you want to change it or not

maltesander committed:
Adapted to PR review.
